Add summary of comments from Issue #73 (#82)
llvm/llvm-iwg: rfc_code_review/comment_summary.md

# RFC: Code Review Comments Summary

## GitHub Missing Features

* Intuitive Review Comments:
  * While this feedback may be somewhat subjective and depend on personal preferences,
    it was a common issue that was mentioned by many community members.  For example,
    some comments disappear when a patch is updated, sometimes comments are auto-hidden
    and can be easy to miss.  And it can be difficult to distinguish between general

>     review comments and comments on specific lines of code

Author: There wasn't agreements on all these points in the issues, so I tried to give a general summary of what was discussed.  My goal was to emphasize the overall sentiment that Phabricator comments were more intuitive that GitHub.

>     review comments and comments on specific lines of code

Reviewer: I still believe that this paragraph does not reflect appropriately the thread, but I can make a direct edit if you prefer?

>     review comments and comments on specific lines of code

Author: > I still believe that this paragraph does not reflect appropriately the thread, but I can make a direct edit if you prefer?

Yeah, if you want make direct edits or suggestions with that exact text you think would be better that would help a lot.

* Automatic Subscription Rules (Herald Rules)
  * There is no builtin way for people to automatically subscribe to a pull request covering
    a specific part of the codebase.  This is seen as a major downside of GitHub pull requests.
* Configurable Notifications.
  * Configuration for notifications is not very flexible and it can be difficult to manage especially

>      for contributors that subscribe to multiple projects.

Author: The second category is more for what happens to people who aren't subscribed to a pull requests.

>      for contributors that subscribe to multiple projects.

Reviewer: Sorry, I still don't quite get it, can you elaborate?

>      for contributors that subscribe to multiple projects.

Reviewer: Ping on this?

>      for contributors that subscribe to multiple projects.

Author: I deleted the "Configurable Notifications" bullet point and kept the one that mentioned "Herald Rules".

* Good support for "Stack Reviews".
  * "Stack Reviews" are reviews of multiple dependant patches.  This is something that is easy to do

>      with Phabricator, but difficult to do with GitHub.

Author: As someone with very little experience with pull requests, this was my take away from the overall discussion.   There was a lot of back and forth and different ideas, and for me it wasn't clear which one was right.  I also am trying to consider that the audience of this summary may also be people who are unfamiliar with pull requests, so I don't know if there is a lot of value in listing the various solutions for this.

>      with Phabricator, but difficult to do with GitHub.

Reviewer: Instead of "difficult to do", it seems more accurate to present it as "does not have an appropriate equivalent in GitHub".

## GitHub Positive Featues

* Familiar for new contributors.
  * Phabricator is seen by some as a barrier of entry for new contributors.
* Can enforce developer policy / restrict pushes.

>   * With GitHub pull requests we can enforce parts of the developer policy with automated checks.

Author: What's a better way to phrases this ?  s/Can/We could if we wanted to/

>   * With GitHub pull requests we can enforce parts of the developer policy with automated checks.

Reviewer: This can't be in a "GitHub positive Features" IMO: because this isn't part of the A/B comparison of the review tooling **within the bounds of our process**.

You may want to add another specific section like `## Others` at the end to add something like that indeed "If we wanted to change our workflow in the future, Pull-Requests would allow us to enforce that code contributions are going through mandatory automated checks" or something like that.

* Third-party hosting and support
  * Phabricator is no longer maintained upstream.
* Easier to integrate with CI and other kinds of automation.
  * GitHub actions makes it very easy to automate parts of our review process.
  * One point that was made in the comments is that it is hard to have reliable pre-commit testing with Phabricator,
    because Phabricator deals only with patches, so if the patch doesn't apply on the current main branch or some
    other known commit then you can't run CI on the patch.  With GitHub pull requests everything is a branch, so
    you can always run CI against them even if the patch is a little bit out-of-date with main.
* Merge from UI
  * You can merge pull requests directly from the web UI, which is much more simple than the Phabricator
    process of manually (or with arc) downloading the patch, applying it and pushing.  This is both a time

>     saver and more intuitive for new contributors.

Reviewer 2: If we have PRs and pre-merge testing, i would like to have an auto-submit option:
* The user sets a flag for auto-merge (e.g. via a comment or label)
* The PR is auto-merged, once some agreed conditions are met (e.g. pre-merge checks passing, at least n human approvals).

This would reduce the friction of pre-merge testing, especially for small patches.

>     saver and more intuitive for new contributors.

Reviewer: It's not clear in what you're saying that this helps new contributors who don't have commit access and so can't trigger the merge (neither through a button, and hopefully not with a comment!).
Also there a bunch of things up in the air in your answer: for example what will the commit message looks like if a PR is auto-merged?

> This would reduce the friction of pre-merge testing, especially for small patches.

I agree, I'm very supportive of a "fire-and-forget" submit queue being available. But that's not the claim here: there not mention of **testing** in this paragraph.

>     saver and more intuitive for new contributors.

Author: >How is it "more intuitive for new contributors"? Since they can't merge the pull-request themselves anyway and will need someone to do it for them, it should be exactly identical?

This was mentioned in the comments on the issue.

>     saver and more intuitive for new contributors.

Reviewer: That does not seem enough of a reason to me to just phrase it this way.

* IDE Integration
  * Some IDEs have GitHub PR integration which can simplify the process of creating a pull request.
* Rich diffs of .rst files.
  * GitHub will render .rst files which makes them much easier to reivew.
* Alternative interfaces  (https://pullrequest.com, reviewable.io)
  * There are alternative interfaces that layer on top of GitHub Pull Requests that provide additional
    functionality.

> 

Reviewer: ```suggestion
* E-mails don't include patch content; 
```

We rely on this for having the mailing list archives as a system of record. This was even a condition to accept Phabricator in the first place.

Reviewer: ```suggestion
* Not a single workflow: many options to do the same thing (for example: rebase+force-push or not?)
```

Not sure how to phrase it, but this is something that makes the overall list of missing feature a bit hard. Because for example we may say that GitHub does not support "Incremental review by looking at only the delta", but someone may say that you can -- if you never ever force-push. However in this case you can't rebase and need to merge main into the PR branch? 

So there is a list of things that we possibly won't have because we'll want to choose a specific workflow and our choice will make some things possible but others harder.

Author: Are these suggestions for the Missing Features section  ?

### GitHub Pull Request Recommended Action Items

* Write a guide for how to create "Stacked Reivews".
* Decide if we allow Pull Request branches in the main repo.
* Define a policy for merging pull requests with multiple commits.
* Determine how can we emulate Herald Rules.
* Write a guide for filtering notifications.

> * Investigate alternative interfaces (https://pullrequest.com, reviewable.io)

Reviewer 2: ```suggestion
* Investigate alternative interfaces (https://pullrequest.com, reviewable.io)
* Start discussion around pre-merge testing and what it would take to make it mandatory
```

Author: I think we should be doing this last point anyway even if we aren't moving to GitHub PRs.  I don't know if I want to link it to this.

# RFC: Code Review Comments Summary

## GitHub Missing Features

* UI Deficiencies: 

>    * On a subjective level, many community members mentioned they are less

Reviewer: ```suggestion
   * On a subjective level, some community members mentioned they are less
```

      reporting that they are using separate browser tabs to review on GitHub.
   * GitHub does not allow to comment on arbitrary line outside of the
      immediate context of the changed lines.
   * Comments become hidden and discussion marked as "obsolete"
      automatically, even when the comment isn't actually addressed or
      the discussion isn't marked "resolved" by the author. In Phabricator
      hiding is an explicit action from each individual.
      This leads to a recurring complain of an impression of "disappearing
      comments": they disappear from the diff view, but remain in the
      history of the "Conversation" tab, disconnected from the patch.
   * In case of force-push (either amend or rebase):
     * the UI does not provide a direct access to view the changes since a
        previous state.
     * Comments can't be viewed in their original context, you only see the
        4 lines above the comment.
* Automatic Subscription Rules (Herald Rules)
  * There is no builtin way for people to automatically subscribe to a pull request covering
    a specific part of the codebase.  This is seen as a major downside of GitHub pull requests.

> * Configurable Notifications.
>   * Configuration for notifications is not very flexible and it can be difficult to manage especially
>      for contributors that subscribe to multiple projects.

Author: ```suggestion
```

* Good support for "Stack Reviews".
  * "Stack Reviews" are reviews of multiple dependant patches.  This is something that is easy to do
     with Phabricator, but does not have an equivalent with GitHub.

> * GitHub isn't open-source and does not offer us any way to implement missing features of customize it for the needs of the project.
> ## GitHub Positive Features

Reviewer: ```suggestion
* GitHub isn't open-source and does not offer us any way to implement missing features of customize it for the needs of the project.

## GitHub Positive Features
```

* Familiar for new contributors.
  * Phabricator is seen by some as a barrier of entry for new contributors.
* Can enforce developer policy / restrict pushes.
  * With GitHub pull requests we can enforce parts of the developer policy with automated checks.
* Third-party hosting and support
  * Phabricator is no longer maintained upstream.
* Easier to integrate with CI and other kinds of automation.
  * GitHub actions makes it very easy to automate parts of our review process.
  * One point that was made in the comments is that it is hard to have reliable pre-commit testing with Phabricator,
    because Phabricator deals only with patches, so if the patch doesn't apply on the current main branch or some
    other known commit then you can't run CI on the patch.  With GitHub pull requests everything is a branch, so
    you can always run CI against them even if the patch is a little bit out-of-date with main.
* Merge from UI

>   * You can merge pull requests directly from the web UI, which is much more simple than the Phabricator
>     process of manually (or with arc) downloading the patch, applying it and pushing.  This is both a time
>     saver and more intuitive for new contributors.

Reviewer: ```suggestion
  * You can merge pull requests directly from the web UI, which avoid leaving the review tool
     to push to main, avoid having to manipulate command line tools (`arc` or `patch`, `git`).
     Phabricator does not have an equivalent.
```

* IDE Integration
  * Some IDEs have GitHub PR integration which can simplify the process of creating a pull request.
* Rich diffs of .rst files.
  * GitHub will render .rst files which makes them much easier to reivew.
* Alternative interfaces  (https://pullrequest.com, reviewable.io)
  * There are alternative interfaces that layer on top of GitHub Pull Requests that provide additional
    functionality.

> * E-mails don't include patch content; 

Reviewer: ```suggestion

```

> ### GitHub Pull Request Recommended Action Items

Reviewer: ```suggestion

### GitHub Pull Request Recommended Action Items
```

      hiding is an explicit action from each individual.
      This leads to a recurring complain of an impression of "disappearing
      comments": they disappear from the diff view, but remain in the
      history of the "Conversation" tab, disconnected from the patch.
   * In case of force-push (either amend or rebase):
     * the UI does not provide a direct access to view the changes since a
        previous state.
     * Comments can't be viewed in their original context, you only see the
        4 lines above the comment.
* Automatic Subscription Rules (Herald Rules)
  * There is no builtin way for people to automatically subscribe to a pull request covering
    a specific part of the codebase.  This is seen as a major downside of GitHub pull requests.
* Configurable Notifications.
  * Configuration for notifications is not very flexible and it can be difficult to manage especially
     for contributors that subscribe to multiple projects.
* Good support for "Stack Reviews".
  * "Stack Reviews" are reviews of multiple dependant patches.  This is something that is easy to do
     with Phabricator, but does not have an equivalent with GitHub.

> * GitHub isn't open-source and does not offer us any way to implement missing features of customize it for the needs of the project.

Reviewer: * E-mails don't include patch content.

> * GitHub isn't open-source and does not offer us any way to implement missing features of customize it for the needs of the project.

Reviewer: ```suggestion
* GitHub isn't open-source and does not offer us any way to implement missing features of customize it for the needs of the project.
* E-mails don't include patch content; 
* ```

  * GitHub actions makes it very easy to automate parts of our review process.
  * One point that was made in the comments is that it is hard to have reliable pre-commit testing with Phabricator,
    because Phabricator deals only with patches, so if the patch doesn't apply on the current main branch or some
    other known commit then you can't run CI on the patch.  With GitHub pull requests everything is a branch, so
    you can always run CI against them even if the patch is a little bit out-of-date with main.
* Merge from UI
  * You can merge pull requests directly from the web UI, which avoid leaving the review tool
     to push to main, avoid having to manipulate command line tools (`arc` or `patch`, `git`).
     Phabricator does not have an equivalent.
* IDE Integration
  * Some IDEs have GitHub PR integration which can simplify the process of creating a pull request.
* Rich diffs of .rst files.
  * GitHub will render .rst files which makes them much easier to reivew.
* Alternative interfaces  (https://pullrequest.com, reviewable.io)
  * There are alternative interfaces that layer on top of GitHub Pull Requests that provide additional
    functionality.

> 

Reviewer: ```suggestion

### Workflow Opportunities with GitHub

* Can enforce developer policy / restrict pushes.
   * With GitHub pull requests we can enforce parts of the developer policy with automated checks.
* One point that was made in the comments is that it is hard to have reliable pre-commit testing with Phabricator,
     because Phabricator deals only with patches, so if the patch doesn't apply on the current main branch or some
     other known commit then you can't run CI on the patch.  With GitHub pull requests everything is a branch, so
     you can always run CI against them even if the patch is a little bit out-of-date with main.
```

# RFC: Code Review Comments Summary

## GitHub Missing Features

> * UI Deficiencies: 

Reviewer: ```suggestion
* UI Deficiencies:
```

>    * On a subjective level, some community members mentioned they are less

Reviewer: ```suggestion
  * On a subjective level, some community members mentioned they are less
```

      productive reviewing on GitHub. Using the two tabs between reviewing files
      and discussion comments instead of a unified view lead to some people
      reporting that they are using separate browser tabs to review on GitHub.

>    * GitHub does not allow to comment on arbitrary line outside of the

Reviewer: ```suggestion
  * GitHub does not allow to comment on arbitrary line outside of the
```

>       immediate context of the changed lines.
>    * Comments become hidden and discussion marked as "obsolete"
>       automatically, even when the comment isn't actually addressed or
>       the discussion isn't marked "resolved" by the author. In Phabricator
>       hiding is an explicit action from each individual.
>       This leads to a recurring complain of an impression of "disappearing
>       comments": they disappear from the diff view, but remain in the
>       history of the "Conversation" tab, disconnected from the patch.
>    * In case of force-push (either amend or rebase):
>      * the UI does not provide a direct access to view the changes since a

Reviewer: ```suggestion
      immediate context of the changed lines.
  * Comments become hidden and discussion marked as "obsolete"
      automatically, even when the comment isn't actually addressed or
      the discussion isn't marked "resolved" by the author. In Phabricator
      hiding is an explicit action from each individual.
      This leads to a recurring complain of an impression of "disappearing
      comments": they disappear from the diff view, but remain in the
      history of the "Conversation" tab, disconnected from the patch.
  * In case of force-push (either amend or rebase):
    * the UI does not provide a direct access to view the changes since a
```

        previous state.

>      * Comments can't be viewed in their original context, you only see the

Reviewer: ```suggestion
    * Comments can't be viewed in their original context, you only see the
```

>      * Comments can't be viewed in their original context, you only see the

Reviewer: ```suggestion
    * the UI does not provide a direct access to view the changes since a
        previous state.
    * Comments can't be viewed in their original context, you only see the
```

        4 lines above the comment.
* Automatic Subscription Rules (Herald Rules)
  * There is no builtin way for people to automatically subscribe to a pull request covering
    a specific part of the codebase.  This is seen as a major downside of GitHub pull requests.
* Good support for "Stack Reviews".
  * "Stack Reviews" are reviews of multiple dependant patches.  This is something that is easy to do
     with Phabricator, but does not have an equivalent with GitHub.
* GitHub isn't open-source and does not offer us any way to implement missing features of customize it for the needs of the project.

> * E-mails don't include patch content; 

Reviewer: ```suggestion
* E-mails don't include patch content.
```

     with Phabricator, but does not have an equivalent with GitHub.
* GitHub isn't open-source and does not offer us any way to implement missing features of customize it for the needs of the project.
* E-mails don't include patch content; 

> * ```

Reviewer: ```suggestion
```

  * Phabricator is seen by some as a barrier of entry for new contributors.
* Third-party hosting and support
  * Phabricator is no longer maintained upstream.
* Easier to integrate with CI and other kinds of automation.
  * GitHub actions makes it very easy to automate parts of our review process.
* Merge from UI
  * You can merge pull requests directly from the web UI, which avoid leaving the review tool
     to push to main, avoid having to manipulate command line tools (`arc` or `patch`, `git`).
     Phabricator does not have an equivalent.
* IDE Integration
  * Some IDEs have GitHub PR integration which can simplify the process of creating a pull request.
* Rich diffs of .rst files.
  * GitHub will render .rst files which makes them much easier to reivew.
* Alternative interfaces  (https://pullrequest.com, reviewable.io)
  * There are alternative interfaces that layer on top of GitHub Pull Requests that provide additional
    functionality.

### Workflow Improvement Opportunities with GitHub

* Can enforce developer policy / restrict pushes.

>    * With GitHub pull requests we can enforce parts of the developer policy with automated checks.

Reviewer: ```suggestion
    * With GitHub pull requests we can enforce parts of the developer policy with automated checks.
```

     * the UI does not provide a direct access to view the changes since a
        previous state.
     * Comments can't be viewed in their original context, you only see the
        4 lines above the comment.
* Automatic Subscription Rules (Herald Rules)
  * There is no builtin way for people to automatically subscribe to a pull request covering
    a specific part of the codebase.  This is seen as a major downside of GitHub pull requests.
* Good support for "Stack Reviews".
  * "Stack Reviews" are reviews of multiple dependant patches.  This is something that is easy to do
     with Phabricator, but does not have an equivalent with GitHub.
* GitHub isn't open-source and does not offer us any way to implement missing features of customize it for the needs of the project.
* E-mails don't include patch content; 

## GitHub Positive Features

* Familiar for new contributors.
  * Phabricator is seen by some as a barrier of entry for new contributors.
* Third-party hosting and support
  * Phabricator is no longer maintained upstream.
* Easier to integrate with CI and other kinds of automation.
  * GitHub actions makes it very easy to automate parts of our review process.

>   * One point that was made in the comments is that it is hard to have reliable pre-commit testing with Phabricator,
>     because Phabricator deals only with patches, so if the patch doesn't apply on the current main branch or some
>     other known commit then you can't run CI on the patch.  With GitHub pull requests everything is a branch, so
>     you can always run CI against them even if the patch is a little bit out-of-date with main.

Reviewer: ```suggestion
```

      comments": they disappear from the diff view, but remain in the
      history of the "Conversation" tab, disconnected from the patch.
   * In case of force-push (either amend or rebase):
     * the UI does not provide a direct access to view the changes since a
        previous state.
     * Comments can't be viewed in their original context, you only see the
        4 lines above the comment.
* Automatic Subscription Rules (Herald Rules)
  * There is no builtin way for people to automatically subscribe to a pull request covering
    a specific part of the codebase.  This is seen as a major downside of GitHub pull requests.
* Good support for "Stack Reviews".
  * "Stack Reviews" are reviews of multiple dependant patches.  This is something that is easy to do
     with Phabricator, but does not have an equivalent with GitHub.
* GitHub isn't open-source and does not offer us any way to implement missing features of customize it for the needs of the project.
* E-mails don't include patch content; 

## GitHub Positive Features

* Familiar for new contributors.
  * Phabricator is seen by some as a barrier of entry for new contributors.

> * Can enforce developer policy / restrict pushes.
>   * With GitHub pull requests we can enforce parts of the developer policy with automated checks.

Reviewer: ```suggestion
```
